three-cards-divination/src/test/java/com/cyber/mysticism/tarot/controller/UserControllerFunctionalTest.java

```diff
@@ -0,0 +1,88 @@
+package com.cyber.mysticism.tarot.controller;
+
+import com.cyber.mysticism.tarot.model.TarotUser;
+import com.cyber.mysticism.tarot.service.UserService;
+import com.cyber.mysticism.tarot.service.exceptions.UserAlreadyExistsException;
+import com.cyber.mysticism.tarot.service.exceptions.UserNotFoundException;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectWriter;
+import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.boot.test.mock.mockito.MockBean;
+import org.springframework.http.MediaType;
+import org.springframework.test.web.servlet.MockMvc;
+
+import java.util.List;
+
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.Mockito.*;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest
+@AutoConfigureMockMvc
+class UserControllerFunctionalTest {
+
+    private static final TarotUser user;
+
+    static {
+        user = new TarotUser();
+        user.setUsername("user");
+        user.setEmail("email");
+    }
```

Comment:
**Suggestion:** Consider using `@BeforeEach` to setup common test data to avoid code duplication. [best practice]
```suggestion
    @BeforeEach
    void setup() {
        user = new TarotUser();
        user.setUsername("user");
        user.setEmail("email");
    }
```